enzyme/include/enzyme/utils

```diff
@@ -43,37 +43,67 @@ namespace enzyme {
     struct Active{
       T value;
       Active(const T& v) : value(v) {};
-      Active(T&& v) : value(v) {};
+      Active(T&& v) : value(std::move(v)) {};
       Active(const Active<T>&) = default;
       Active(Active<T>&&) = default;
       operator T&() { return value; }
     };
 
+    namespace detail {
+
+        /**
+         * @brief Helper base for all `Duplicated`-type arguments
+         *
+         * Due to essentially the same implementation, it prevents code
+         * duplication
+         */
+        template < typename T >
+        struct DupArg {
+            // Argument as regular value. Also works for pointer types
+            T value;
+            T shadow;
+            DupArg(const T& v, const T& s) : value(v), shadow(s) {};
+            DupArg(T&& v, T&& s) : value(v), shadow(s) {};
+            DupArg(const DupArg<T>&) = default;
+            DupArg(DupArg<T>&&) = default;
+
+            __attribute__((always_inline))
+            T valueExp() const { return this->value; }
+            __attribute__((always_inline))
+            T shadowExp() const { return this->shadow; }
+        };
+
+        template < typename T >
+        struct DupArg<T&> {
+            T& value;
+            T& shadow;
+            DupArg(T& v, T& s) : value(v), shadow(s) {}
+            DupArg(const DupArg<T&>&) = default;
+            DupArg(DupArg<T&>&&) = default;
+
+            __attribute__((always_inline))
+            const T* valueExp() const { return &this->value; }
```

Comment:
This shouldn't be const here, no?

Reply:
It you want to keep const correctness here, maybe we should store T* for value and shadow?

Reply:
Yeah `T* value` would be much simpler as it would do the same thing that `valueExp` and `shadowExp` are doing anyway...

As for the `const`... I believe it doesn't matter a lot at this point where the `const` is because the next destination for these pointers is `__enzyme_autodiff` and by the time it reaches IR the semantic value of `const` will be already lost...

Reply:
That's true, but if Duplicated gets used semantically for how we define custom c++ derivative rules we want users to be able to modify things appropriately.

Reply:
That would be a bit tricky to do. Let's take `enzyme::Duplicated<const float&>` for example. What should it's `shadow` type be? In reverse mode it will get modified by the AD, so its const-correct type is `float&`. On the contrary, forward mode will only read the value from that address, so the user should be able to bind a `const` to that reference, which means that the shadow in this case needs to be a `const float&`.

Unless I'm missing something here, there is a contradiction here in what `Reverse` and `Forward` mode want from the shadow parameter in `Duplicated` which can't be solved unless there is a separate class for the `Forward` mode specifically

Reply:
Yeah we really should fork off duplicated in forward mode to just be Dual and have the semantic changes.

In any case yeah, we may need to remove the const from the primal type in the shadow.

This doesn’t need to stop the PR from getting in to begin with (already improving tuings) — but something to follow up on